### PR TITLE
Added initial cluster support from redis-cluster-rs crate

### PR DESCRIPTION
**Pull request by mitsuhiko**
This PR adds cluster support.  Right now it's based on the redis-cluster-rs crate but
larger changes are necessary for this to work better in practice.  In
particular I am going to port the routing code from the python redis cluster
crate which is more correct than what is in the code right now.

Missing:

* [x] fanouts for `AllNodes`
* [x] cluster stores all slots individually in the hashmap
* [x] ~~implement `MergeResults` for `Vec<Value>`~~
* [x] fix travis tests
* [x] add new error types instead of using `ExtensionError`.
* [x] added basic ASKING logic

For later:

* fanouts for `AllMasters`
* add a code path where commands do not need to be unparsed

I decided to not implement pipelining for now. There are bigger issues with the original library this is based on I need to sort out first.

~~I asked the original author if BSD relicensing is okay.~~

Update: I got an okay for the relicensing.

### CODE REVIEW

**Comment by mitsuhiko** [comment]
This changes the license to MIT as well because that's what the cluster code's license is. I sent an email out to the developers of that crate to establish a relationship but worst case I would consider relicensing this crate to MIT to match.

**Comment by Marwes** [comment]
For async code I ported `redis-cluster-rs` in https://github.com/Marwes/redis-cluster-rs . If this ends up being merged we probably want to merge that in as well.

There are some async independent fixes in there as well, such as fixing the key extraction https://github.com/mitsuhiko/redis-rs/pull/239/files#diff-7244e27d2ca38bc882a487eef8b39499R566 and avoding to allocate a 16384 element HashMap for the slot map.

**Comment by mitsuhiko** [comment]
@Marwes oh neat. Didn't see that. I will have a look. They key extraction i wanted / have to change anyways because it does not route `EVALSHA` and others correctly. I will see what I can use from there (probably most).

**Comment by Marwes** [comment]
We probably want to change the `ConnectionLike` trait as `[u8]` is really to low level for what clustering clients needs.

**Comment by mitsuhiko** [comment]
@Marwes agreed. Though I'm not sure where to best put the abstraction. Some commands need fanouts, others need to go to random nodes etc. At this point it might even make sense to not pass commands as vectors around but as enums and attach methods to it.

**Comment by mitsuhiko** [comment]
@Marwes I think it would make more sense to based at least the fanout logic on tokio at all times internally. That way those queries can be sent in parallel even if sync mode is used.

File: src/cluster.rs [hunk 1]
```diff
@@ -0,0 +1,916 @@
...
+    /// # Errors
+    ///
+    /// If it is failed to open connections and to create slots, an error is returned.
+    pub fn get_connection(&self) -> RedisResult<ClusterConnection> {
+        ClusterConnection::new(
+            self.initial_nodes.clone(),
+            self.readonly,
+            self.password.clone(),
+        )
+    }
```
Comment (RicoGit) [inline src/cluster.rs]: Does "get_async_connection" appear in next PRs?

**Comment by untitaker** [comment]
What do we do with this? We use this extensively already.

**Comment by mitsuhiko** [comment]
I will try to get this merged.

**Comment by mitsuhiko** [comment]
Will merge this today.

**Pull request merged**